src/enable_variable_ops/ngraph_replace_variable_modifiers.cc

```diff
@@ -0,0 +1,210 @@
+/*******************************************************************************
+ * Copyright 2019-2020 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/graph/node_builder.h"
+
+#include "ngraph_api.h"
+#include "ngraph_capture_variables.h"
+#include "ngraph_replace_variable_modifiers.h"
+#include "ngraph_utils.h"
+
+using namespace std;
+namespace ng = ngraph;
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+Status ReplaceModifiers(Graph* graph, int graph_id) {
+  // Go over the nodes and replace variable modifiers
+  // Each Modifier is replaced with the corresponding computational TF
+  // graph followed by NGraphAssign Op
+  // If there is an incoming control edge to the Modifier Op
+  // It is attached to the first op in the series of the computation TF graph
+  for (auto node : graph->op_nodes()) {
+    if (node->type_string() == "NGraphAssignSub" ||
+        node->type_string() == "NGraphAssignAdd") {
+      NodeBuilder::NodeOut input_ref;
+      NodeBuilder::NodeOut input_val;
+
+      DataType dtype;
+      TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "T", &dtype));
+
+      for (auto edge : node->in_edges()) {
+        if (edge == NULL) {
+          continue;
+        }
+        if (edge->dst()->IsOp() && !edge->IsControlEdge() &&
+            IsRefType(edge->dst()->input_type(edge->dst_input()))) {
+          input_ref = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+        } else {
+          input_val = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+        }
+      }
+
+      string op_type =
+          (node->type_string() == "NGraphAssignSub") ? "Sub" : "Add";
+      string op_name_suffix = "_" + op_type;
+      Node* compute_op;
+      string new_name_sub = node->name() + op_name_suffix;
```

Comment:
A set of unique strings will remain unique only if ALL of them are transformed by adding the same suffix/prefix, but not if only some of them are changed.

Lets say original graph had 2 nodes, `hello` (of type assignadd) and `hello_add` (of type convolution, so we will never change its name)

After this modification, `hello` becomes `hello_add`.

Therefore the new graph has 2 nodes both called `hello_add`

It is quite possible that TF will itself correct the same node names by renaming one of the `hello_add`s to `hello_add_1` or something, or TF might barf. But I am not sure of TF's behaviour in case we force non-unique node names. If TF self corrects we do not have to attach unique names, which has the advantage of giving the node a human-readable name. If TF crashes, then we might have to rely on `NewName`, which might produce an uninterpretable (but unique) name.

Reply:
Yes, I realized it later that it applies when renaming all nodes and not some. I can check TFs behavior whether it throws an error or not. We have been attaching suffix to node names (just_looking) and also naming the encapsulate node without really checking whether it's unique.  By adding the suffix , we add more information to the node name, which can be helpful in certain debugging cases and getting an overall sense of graph transformations that we do at the bridge.

Reply:
what if we attach a unique id(int) to the "name + suffix" if it makes sense?

Reply:
To be addressed in a separate PR. I have added TODO in ngraph_replace_variable_modifiers.cc, that's where we set new names first time.